File: python/src/pi_mono/agent/harness/compaction/branch_summarization.py
"""Branch summarization for agent harness."""

from __future__ import annotations

from typing import Any

from pi_mono.ai.stream import complete_simple
from pi_mono.agent.harness.messages import (
    create_branch_summary_message,
    create_compaction_summary_message,
    create_custom_message,
    convert_to_llm,
)
from pi_mono.agent.harness.session.session import Session
from pi_mono.agent.harness.compaction.utils import (
    compute_file_lists,
    create_file_ops,
    extract_file_ops_from_message,
    format_file_operations,
    serialize_conversation,
)
from pi_mono.agent.harness.compaction.compaction import estimate_tokens


NL = "\n"


def _get_entry_field(entry: Any, field_name: str) -> Any:
    if isinstance(entry, dict):
        return entry.get(field_name)
    attr_map = {
        "parentId": "parent_id",
        "parent_id": "parent_id",
        "firstKeptEntryId": "first_kept_entry_id",
        "first_kept_entry_id": "first_kept_entry_id",
        "customType": "custom_type",
        "custom_type": "custom_type",
        "fromId": "from_id",
        "from_id": "from_id",
        "tokensBefore": "tokens_before",
        "tokens_before": "tokens_before",
        "fromHook": "from_hook",
        "from_hook": "from_hook",
        "thinkingLevel": "thinking_level",
        "thinking_level": "thinking_level",
        "activeToolNames": "active_tool_names",
        "active_tool_names": "active_tool_names",
        "modelId": "model_id",
        "model_id": "model_id",
    }
    attr_name = attr_map.get(field_name, field_name)
    return getattr(entry, attr_name, None)


def get_message_from_entry(entry: Any) -> dict | None:
    entry_type = _get_entry_field(entry, "type")
    if entry_type == "message":
        msg = _get_entry_field(entry, "message")
        if msg and msg.get("role") == "toolResult":
            return None
        return msg
    if entry_type == "custom_message":
        return create_custom_message(
            _get_entry_field(entry, "customType"),
            _get_entry_field(entry, "content"),
            _get_entry_field(entry, "display"),
            _get_entry_field(entry, "details"),
            _get_entry_field(entry, "timestamp"),
        )
    if entry_type == "branch_summary":
        return create_branch_summary_message(
            _get_entry_field(entry, "summary"),
            _get_entry_field(entry, "fromId"),
            _get_entry_field(entry, "timestamp"),
        )
    if entry_type == "compaction":
        return create_compaction_summary_message(
            _get_entry_field(entry, "summary"),
            _get_entry_field(entry, "tokensBefore"),
            _get_entry_field(entry, "timestamp"),
        )
    return None


def prepare_branch_entries(entries: list[Any], token_budget: int = 0) -> dict[str, Any]:
    messages: list[dict] = []
    file_ops = create_file_ops()
    total_tokens = 0

    for entry in entries:
        if (
            _get_entry_field(entry, "type") == "branch_summary"
            and not _get_entry_field(entry, "fromHook")
            and _get_entry_field(entry, "details")
        ):
            details = _get_entry_field(entry, "details")
            if isinstance(details.get("readFiles"), list):
                for f in details["readFiles"]:
                    file_ops.read.add(f)
            if isinstance(details.get("modifiedFiles"), list):
                for f in details["modifiedFiles"]:
                    file_ops.edited.add(f)

    for i in range(len(entries) - 1, -1, -1):
        entry = entries[i]
        message = get_message_from_entry(entry)
        if not message:
            continue
        extract_file_ops_from_message(message, file_ops)

        tokens = estimate_tokens(message)
        if token_budget > 0 and total_tokens + tokens > token_budget:
            entry_type = _get_entry_field(entry, "type")
            if entry_type in ("compaction", "branch_summary"):
                if total_tokens < token_budget * 0.9:
                    messages.insert(0, message)
                    total_tokens += tokens
            break

        messages.insert(0, message)
        total_tokens += tokens

    return {"messages": messages, "fileOps": file_ops, "totalTokens": total_tokens}


BRANCH_SUMMARY_PREAMBLE = "The user explored a different conversation branch before returning here.\nSummary of that exploration:\n\n"

BRANCH_SUMMARY_PROMPT = """Create a structured summary of this conversation branch for context when returning later.

Use this EXACT format:

## Goal
[What was the user trying to accomplish in this branch?]

## Constraints & Preferences
- [Any constraints, preferences, or requirements mentioned]
- [Or "(none)" if none were mentioned]

## Progress
### Done
- [x] [Completed tasks/changes]

### In Progress
- [ ] [Work that was started but not finished]

### Blocked
- [Issues preventing progress, if any]

## Key Decisions
- **[Decision]**: [Brief rationale]

## Next Steps
1. [What should happen next to continue this work]

Keep each section concise. Preserve exact file paths, function names, and error messages."""


SUMMARIZATION_SYSTEM_PROMPT = """You are a context summarization assistant. Your task is to read a conversation between a user and an AI coding assistant, then produce a structured summary following the exact format specified.

Do NOT continue the conversation. Do NOT respond to any questions in the conversation. ONLY output the structured summary."""


async def collect_entries_for_branch_summary(
    session: Session,
    old_leaf_id: str | None,
    target_id: str,
) -> dict[str, Any]:
    if not old_leaf_id:
        return {"entries": [], "commonAncestorId": None}

    old_path_set = {e.id for e in await session.get_branch(old_leaf_id)}
    target_path = await session.get_branch(target_id)
    common_ancestor_id: str | None = None

    for i in range(len(target_path) - 1, -1, -1):
        if target_path[i].id in old_path_set:
            common_ancestor_id = target_path[i].id
            break

    entries: list[dict] = []
    current: str | None = old_leaf_id

    while current and current != common_ancestor_id:
        entry = await session.get_entry(current)
        if not entry:
            raise Exception(f"Entry {current} not found")
        entries.append(entry)
        current = entry.parent_id

    entries.reverse()
    return {"entries": entries, "commonAncestorId": common_ancestor_id}


async def generate_branch_summary(
    entries: list[dict],
    options: dict[str, Any],
) -> dict[str, Any]:
    from pi_mono.agent.harness.compaction.compaction import Result as CompactionResult

    model = options["model"]
    api_key = options["apiKey"]
    headers = options.get("headers")
    signal = options.get("signal")
    custom_instructions = options.get("customInstructions")
    replace_instructions = options.get("replaceInstructions", False)
    reserve_tokens = options.get("reserveTokens", 16384)

    context_window = model.get("contextWindow", 128000)
    token_budget = context_window - reserve_tokens

    prep = prepare_branch_entries(entries, token_budget)
    messages = prep["messages"]
    file_ops = prep["fileOps"]

    if not messages:
        return CompactionResult.ok(
            {
                "summary": "No content to summarize",
                "readFiles": [],
                "modifiedFiles": [],
            }
        )

    llm_messages = convert_to_llm(messages)
    conversation_text = serialize_conversation(llm_messages)

    if replace_instructions and custom_instructions:
        instructions = custom_instructions
    elif custom_instructions:
        instructions = f"{BRANCH_SUMMARY_PROMPT}{NL}{NL}Additional focus: {custom_instructions}"
    else:
        instructions = BRANCH_SUMMARY_PROMPT

    prompt_text = f"<conversation>{NL}{conversation_text}{NL}</conversation>{NL}{NL}{instructions}"

    summarization_messages = [
        {
            "role": "user",
            "content": [{"type": "text", "text": prompt_text}],
            "timestamp": 0,
        }
    ]

    completion_options = {
        "maxTokens": 2048,
        "apiKey": api_key,
        "headers": headers,
        "signal": signal,
    }
    try:
        response = await complete_simple(
            model,
            {"systemPrompt": SUMMARIZATION_SYSTEM_PROMPT, "messages": summarization_messages},
            completion_options,
        )
    except Exception as e:
        return CompactionResult.err(Exception(f"Branch summary failed: {e}"))

    if response.get("stopReason") == "aborted":
        return CompactionResult.err(Exception("Branch summary aborted"))

    if response.get("stopReason") == "error":
        return CompactionResult.err(
            Exception(f"Branch summary failed: {response.get('errorMessage', 'Unknown error')}")
        )

    summary = NL.join(c["text"] for c in response.get("content", []) if c.get("type") == "text")

    summary = BRANCH_SUMMARY_PREAMBLE + summary
    read_files, modified_files = compute_file_lists(file_ops)
    summary += format_file_operations(read_files, modified_files)

    return CompactionResult.ok(
        {
            "summary": summary or "No summary generated",
            "readFiles": read_files,
            "modifiedFiles": modified_files,
        }
    )


def collect_entries_for_branch_summary_sync(
    session_manager: Any,
    old_leaf_id: str | None,
    target_id: str,
) -> dict[str, Any]:
    if not old_leaf_id:
        return {"entries": [], "commonAncestorId": None}

    old_path = session_manager.get_branch(old_leaf_id)
    old_path_set = {entry["id"] for entry in old_path}
    target_path = session_manager.get_branch(target_id)
    common_ancestor_id: str | None = None

    for entry in reversed(target_path):
        if entry["id"] in old_path_set:
            common_ancestor_id = entry["id"]
            break

    entries: list[dict[str, Any]] = []
    current: str | None = old_leaf_id
    while current and current != common_ancestor_id:
        entry = session_manager.get_entry(current)
        if not entry:
            raise RuntimeError(f"Entry {current} not found")
        entries.append(entry)
        current = entry.get("parentId")

    entries.reverse()
    return {"entries": entries, "commonAncestorId": common_ancestor_id}


collectEntriesForBranchSummary = collect_entries_for_branch_summary
generateBranchSummary = generate_branch_summary

File: packages/agent/src/harness/compaction/branch-summarization.ts
import { contentText, type Model, type Models, type RetryCallbacks, type RetryPolicy } from "@earendil-works/pi-ai";

import type { AgentMessage } from "../../types.ts";
import {
	convertToLlm,
	createBranchSummaryMessage,
	createCompactionSummaryMessage,
	createCustomMessage,
} from "../messages.ts";
import type { BranchSummaryResult, Session, SessionTreeEntry } from "../types.ts";
import { BranchSummaryError, err, ok, type Result, SessionError } from "../types.ts";
import { completeSimpleWithRetries, estimateTokens, SUMMARIZATION_SYSTEM_PROMPT } from "./compaction.ts";
import {
	computeFileLists,
	createFileOps,
	extractFileOpsFromMessage,
	type FileOperations,
	formatFileOperations,
	serializeConversation,
} from "./utils.ts";

/** 存储在生成的分支摘要条目上的文件操作详情。 */
export interface BranchSummaryDetails {
	/** 在探索被摘要分支时读取的文件。 */
	readFiles: string[];
	/** 在探索被摘要分支时修改的文件。 */
	modifiedFiles: string[];
}

export type { FileOperations } from "./utils.ts";

/** 准备好用于摘要的分支内容。 */
export interface BranchPreparation {
	/** 选择用于分支摘要的消息。 */
	messages: AgentMessage[];
	/** 从分支提取的文件操作。 */
	fileOps: FileOperations;
	/** 所选消息的预估 token 数。 */
	totalTokens: number;
}

/** 选择用于分支摘要的条目。 */
export interface CollectEntriesResult {
	/** 按时间顺序排列的待摘要条目。 */
	entries: SessionTreeEntry[];
	/** 上一个叶节点和目标条目之间最深层的公共祖先。 */
	commonAncestorId: string | null;
}

/** 生成分支摘要的选项。 */
export interface GenerateBranchSummaryOptions {
	/** 摘要请求经过的提供商集合；拥有认证解析。 */
	models: Models;
	/** 用于摘要的模型。 */
	model: Model<any>;
	/** 摘要请求的中止信号。 */
	signal: AbortSignal;
	/** 追加到或替换默认提示词的可选指令。 */
	customInstructions?: string;
	/** 用自定义指令替换默认提示词，而不是追加。 */
	replaceInstructions?: boolean;
	/** 为提示词和模型输出预留的 token 数。默认为 16384。 */
	reserveTokens?: number;
	/** 可选的重试策略，用于瞬时摘要错误。 */
	retry?: RetryPolicy;
	/** 可选的重试回调。 */
	callbacks?: RetryCallbacks;
}

/** 在导航到不同会话树条目之前收集应被摘要的条目。 */
export async function collectEntriesForBranchSummary(
	session: Session,
	oldLeafId: string | null,
	targetId: string,
): Promise<CollectEntriesResult> {
	if (!oldLeafId) {
		return { entries: [], commonAncestorId: null };
	}
	const oldPath = new Set((await session.getBranch(oldLeafId)).map((e) => e.id));
	const targetPath = await session.getBranch(targetId);
	let commonAncestorId: string | null = null;
	for (let i = targetPath.length - 1; i >= 0; i--) {
		if (oldPath.has(targetPath[i].id)) {
			commonAncestorId = targetPath[i].id;
			break;
		}
	}
	const entries: SessionTreeEntry[] = [];
	let current: string | null = oldLeafId;

	while (current && current !== commonAncestorId) {
		const entry = await session.getEntry(current);
		if (!entry) throw new SessionError("invalid_session", `Entry ${current} not found`);
		entries.push(entry as SessionTreeEntry);
		current = entry.parentId;
	}
	entries.reverse();

	return { entries, commonAncestorId };
}
/**
 * 将会话树条目转换为具体的 {@link AgentMessage}，用于分支摘要。
 *
 * 与此函数的压缩版本不同，工具结果消息被排除，
 * 因为它们增加了噪音而无助于分支摘要的上下文。
 * 压缩条目在此处也被映射为摘要消息。
 */
function getMessageFromEntry(entry: SessionTreeEntry): AgentMessage | undefined {
	switch (entry.type) {
		case "message":
			if (entry.message.role === "toolResult") return undefined;
			return entry.message;

		case "custom_message":
			return createCustomMessage(entry.customType, entry.content, entry.display, entry.details, entry.timestamp);

		case "branch_summary":
			return createBranchSummaryMessage(entry.summary, entry.fromId, entry.timestamp);

		case "compaction":
			return createCompactionSummaryMessage(entry.summary, entry.tokensBefore, entry.timestamp);
		case "thinking_level_change":
		case "model_change":
		case "active_tools_change":
		case "custom":
		case "label":
		case "session_info":
		case "leaf":
			return undefined;
	}
}

/**
 * 在可选的 token 预算内准备用于摘要的分支条目。
 *
 * 按时间倒序遍历条目，收集消息直到 token 预算耗尽。
 * 文件操作从条目级分支摘要详情和单个助手消息中提取。
 * 当设置了预算且即将超出时，压缩和分支摘要条目如果
 * 落在预算的 90% 以内仍会被包含，以保留关键的上下文标记。
 */
export function prepareBranchEntries(entries: SessionTreeEntry[], tokenBudget: number = 0): BranchPreparation {
	const messages: AgentMessage[] = [];
	const fileOps = createFileOps();
	let totalTokens = 0;
	for (const entry of entries) {
		if (entry.type === "branch_summary" && !entry.fromHook && entry.details) {
			const details = entry.details as BranchSummaryDetails;
			if (Array.isArray(details.readFiles)) {
				for (const f of details.readFiles) fileOps.read.add(f);
			}
			if (Array.isArray(details.modifiedFiles)) {
				for (const f of details.modifiedFiles) {
					fileOps.edited.add(f);
				}
			}
		}
	}
	for (let i = entries.length - 1; i >= 0; i--) {
		const entry = entries[i];
		const message = getMessageFromEntry(entry);
		if (!message) continue;
		extractFileOpsFromMessage(message, fileOps);

		const tokens = estimateTokens(message);
		if (tokenBudget > 0 && totalTokens + tokens > tokenBudget) {
			if (entry.type === "compaction" || entry.type === "branch_summary") {
				if (totalTokens < tokenBudget * 0.9) {
					messages.unshift(message);
					totalTokens += tokens;
				}
			}
			break;
		}

		messages.unshift(message);
		totalTokens += tokens;
	}

	return { messages, fileOps, totalTokens };
}

/** 附加在每个分支摘要前面的前言，用于表明它涵盖了不同的对话分支。 */
const BRANCH_SUMMARY_PREAMBLE = `The user explored a different conversation branch before returning here.
Summary of that exploration:

`;

/** 指示模型为已放弃的对话分支创建结构化摘要的提示词模板。 */
const BRANCH_SUMMARY_PROMPT = `Create a structured summary of this conversation branch for context when returning later.

Use this EXACT format:

## Goal
[What was the user trying to accomplish in this branch?]

## Constraints & Preferences
- [Any constraints, preferences, or requirements mentioned]
- [Or "(none)" if none were mentioned]

## Progress
### Done
- [x] [Completed tasks/changes]

### In Progress
- [ ] [Work that was started but not finished]

### Blocked
- [Issues preventing progress, if any]

## Key Decisions
- **[Decision]**: [Brief rationale]

## Next Steps
1. [What should happen next to continue this work]

Keep each section concise. Preserve exact file paths, function names, and error messages.`;

/**
 * 为已放弃的分支条目生成摘要。
 *
 * 从分支条目中准备消息，token 预算由模型的上下文窗口减去预留 token 数得出。
 * 文件操作（读取、写入、编辑）被跟踪并以元数据标签的形式追加。
 * 摘要以一段前言开头，说明用户曾探索了不同的分支。
 */
export async function generateBranchSummary(
	entries: SessionTreeEntry[],
	options: GenerateBranchSummaryOptions,
): Promise<Result<BranchSummaryResult, BranchSummaryError>> {
	const {
		models,
		model,
		signal,
		customInstructions,
		replaceInstructions,
		reserveTokens = 16384,
		retry,
		callbacks,
	} = options;
	const contextWindow = model.contextWindow || 128000;
	const tokenBudget = contextWindow - reserveTokens;

	const { messages, fileOps } = prepareBranchEntries(entries, tokenBudget);

	if (messages.length === 0) {
		return ok({ summary: "No content to summarize", readFiles: [], modifiedFiles: [] });
	}
	const llmMessages = convertToLlm(messages);
	const conversationText = serializeConversation(llmMessages);
	let instructions: string;
	if (replaceInstructions && customInstructions) {
		instructions = customInstructions;
	} else if (customInstructions) {
		instructions = `${BRANCH_SUMMARY_PROMPT}\n\nAdditional focus: ${customInstructions}`;
	} else {
		instructions = BRANCH_SUMMARY_PROMPT;
	}
	const promptText = `<conversation>\n${conversationText}\n</conversation>\n\n${instructions}`;

	const summarizationMessages = [
		{
			role: "user" as const,
			content: [{ type: "text" as const, text: promptText }],
			timestamp: Date.now(),
		},
	];
	const response = await completeSimpleWithRetries(
		models,
		model,
		{ systemPrompt: SUMMARIZATION_SYSTEM_PROMPT, messages: summarizationMessages },
		{ signal, maxTokens: 2048 },
		retry,
		callbacks,
	);
	if (response.stopReason === "aborted") {
		return err(new BranchSummaryError("aborted", response.errorMessage || "Branch summary aborted"));
	}
	if (response.stopReason === "error") {
		return err(
			new BranchSummaryError(
				"summarization_failed",
				`Branch summary failed: ${response.errorMessage || "Unknown error"}`,
			),
		);
	}

	let summary = contentText(response.content);
	summary = BRANCH_SUMMARY_PREAMBLE + summary;
	const { readFiles, modifiedFiles } = computeFileLists(fileOps);
	summary += formatFileOperations(readFiles, modifiedFiles);

	return ok({
		summary: summary || "No summary generated",
		usage: response.usage,
		readFiles,
		modifiedFiles,
	});
}
